src/blr/cli/clusterrmdup.py

```diff
@@ -12,237 +12,297 @@
 
-    bc_new = utils.get_bamtag(pysam_read=read, tag=barcode_tag)
-    if not bc_new:
-        summary["Non tagged reads"] += 2
-        rp_meet_requirements = False
+def update_positions(positions, current_position, read, mate, barcode):
+    """
+    Update positions list with current positions and read pair information.
+    :param positions: list: Postions buffer.
+    :param current_position: tuple: Position information
+    :param read: pysam.AlignedSegment
+    :param mate: pysam.AlignedSeqment
+    :param barcode: str: Barcode string.
+    """
 
-    return rp_meet_requirements, bc_new
+    if current_position in positions:
+        positions[current_position].add_read_pair_and_barcode(read=read, mate=mate, barcode=barcode)
+    else:
+        positions[current_position] = PositionTracker(position=current_position, read=read,
+                                                      mate=mate, barcode=barcode)
 
 
-class CacheReadPairTracker:
+def find_barcode_duplicates(positions, buffer_dup_pos, merge_dict, window, summary):
     """
-    Stores read pairs and keeps track is reads/mates are marked as duplicate for that set of reads.
+    Parse position to check if they are valid duplicate positions. If so start looking for barcode duplicates.
+    :param positions: list: Position to check for duplicates
+    :param merge_dict: dict: Tracks which barcodes shuold be merged .
+    :param buffer_dup_pos: list: Tracks previous duplicate positions and their barcode sets.
+    :param window: int: Max distance allowed between postions to call barcode duplicate.
+    :param summary: dict
     """
+    positions_to_remove = list()
+    for position in positions.keys():
+        tracked_position = positions[position]
+        if tracked_position.valid_duplicate_position():
+            summary["Reads at duplicate position"] += tracked_position.reads
+            seed_duplicates(
+                merge_dict=merge_dict,
+                buffer_dup_pos=buffer_dup_pos,
+                position=tracked_position.position,
+                position_barcodes=tracked_position.barcodes,
+                window=window
+            )
+        elif tracked_position.checked:
+            positions_to_remove.append(position)
+        else:
+            tracked_position.checked = True
 
-    def __init__(self, rp_pos_tuple, chromosome, read, mate, bc):
+    for position in positions_to_remove:
+        del positions[position]
+
+
+class PositionTracker:
+    """
+    Stores read pairs information relatec to a position and keeps track if reads/mates are marked as duplicate
+    for that set of reads.
+    """
 
-        self.position_tuple_ID = rp_pos_tuple
-        self.chromosome = chromosome
-        self.current_reads = dict()
+    def __init__(self, position, read, mate, barcode):
+        self.position = position
+        self.reads = int()
         self.barcodes = set()
+        self.checked = False
+        self.updated = bool()
```

Comment:
```suggestion
        self.contains_reads = bool()
```

Reply:
I think `updated` makes more sense here although it may not be obvious. As the position is only removed when its outside the buffer the same position may be evaluated multiple times. I added this so that once it had been evaluated it would only be evaluated again if more reads had been added. Thus is called `updated`.

Reply:
Ah, I see. Makes sense!